### PR TITLE
autodoc: better indentation handling when rendering source code

### LINKED ISSUE #19293: better handle indentation when rendering source code
Extracted from #19249.

[Example](https://ziglang.org/documentation/master/std/#std.array_hash_map.ArrayHashMapUnmanaged.count)

```zig
pub fn count(self: Self) usize {
            return self.entries.len;
        }
```

Instead it should render like this:

```zig
pub fn count(self: Self) usize {
    return self.entries.len;
}
```

This happens because it is rendering only the function declaration directly from the file, which in fact is indented two levels. The rendering function needs to detect this and chop off the indentation levels.

#### Case 1

Input:
```zig
    fn add(a: u32, b: u32) u32 {
        return a + b;
    }
```
Expected Output:
```zig
fn add(a: u32, b: u32) u32 {
    return a + b;
}
```

#### Case 2

Input: 
```zig
    const a = 1; fn add(a: u32, b: u32) u32 {
        return a + b;
    }
```
Expected Output:
```zig
fn add(a: u32, b: u32) u32 {
    return a + b;
}
```

#### Case 3

Input: 
```zig
    fn add(a: u32, b: u32) u32 {
return a + b;
    }
```
Expected Output:

```zig
fn add(a: u32, b: u32) u32 {
return a + b;
}
```

#### Case 4

Input: 
```zig
    const a = 1; fn add(a: u32, b: u32) u32 {
return a + b;
    }
```
Expected Output:
```zig
fn add(a: u32, b: u32) u32 {
return a + b;
}
```

Note that those last two cases are planned to become syntax errors:
* #35

### PR DESCRIPTION
**Pull request by HydroH**
Closes #19293

### CODE REVIEW

File: lib/docs/wasm/main.zig [hunk 1]
```diff
@@ -938,6 +938,17 @@ fn file_source_html(
 
     var cursor: usize = token_starts[start_token];
 
+    var indent: usize = 0;
+    if (std.mem.lastIndexOfAny(u8, ast.source[0..cursor], "\r\n")) |newline_index| {
```
Comment (mlugg) [inline lib/docs/wasm/main.zig:942]: Canonical Zig style uses LF line endings (`\n`). Using `\n` here is also conveniently compatible with `\r\n`, so replacing this with `"\n"` will handle both line endings in common use today.

File: lib/docs/wasm/main.zig [hunk 2]
```diff
@@ -1188,6 +1199,31 @@ fn file_source_html(
...
+            indent_idx += 1;
+        } else {
+            break;
+        }
+    }
+    return s[indent_idx..];
+}
+
+fn appendUnindented(out: *std.ArrayListUnmanaged(u8), s: []const u8, indent: usize) !void {
+    var it = std.mem.splitAny(u8, s, "\r\n");
```
Comment (mlugg) [inline lib/docs/wasm/main.zig:1215]: Same as above.

**Review by mlugg — CHANGES_REQUESTED** [review]

**Comment by andrewrk** [comment]
Can you share a screenshot to demonstrate your work?

**Comment by HydroH** [comment]
> Can you share a screenshot to demonstrate your work?

Sure. This is a comparison between this pr and master branch rendered docs. I chose a function with nested blocks and comments, which should be a good demonstration.
image
image

**Review by mlugg — APPROVED** [review]
Looks good to me!

I know what (AFAICT) this code ought to do, but could you give a screenshot of an example where some of the code has lower indentation than the start of the block? e.g.
```zig
const S = struct {
    const foo = {};
    test foo {
        const x: [2]u8 =
        // zig fmt: off
.{ 10, 20 };
        // zig fmt: on
        _ = x;
    }
};
```

Assuming I'm reading this right, the test should render like this (which I would consider ideal):
```zig
test foo {
    const x: [2]u8 =
    // zig fmt: off
.{ 10, 20 };
    // zig fmt: on
    _ = x;
}
```

**Comment by HydroH** [comment]
> Looks good to me!
> 
> I know what (AFAICT) this code ought to do, but could you give a screenshot of an example where some of the code has lower indentation than the start of the block? e.g.
> 
> ```zig
> const S = struct {
>     const foo = {};
>     test foo {
>         const x: [2]u8 =
>         // zig fmt: off
> .{ 10, 20 };
>         // zig fmt: on
>         _ = x;
>     }
> };
> ```
> 
> Assuming I'm reading this right, the test should render like this (which I would consider ideal):
> 
> ```zig
> test foo {
>     const x: [2]u8 =
>     // zig fmt: off
> .{ 10, 20 };
>     // zig fmt: on
>     _ = x;
> }
> ```

This is the result for the following code:
```zig
        pub fn testIndent() void {
            const x: [2]u8 =
                // zig fmt: off
.{ 10, 20 };
            // zig fmt: on
            _ = x;
        }
```
image

**Comment by mlugg** [comment]
Brilliant, I'm happy with this then - thanks!

**Pull request merged**